### PR TITLE
[Merged by Bors] - feat: explicit logarithmic bounds on the harmonic numbers

### PR DESCRIPTION
**Pull request by amellendijk**
Prove $\log(n+1) \le H_n \le 1 + \log(n)$ for all natural numbers $n$. 

There is an existing proof that $H_n$ is not an integer which uses p-adics. Since the new result uses some heavy machinery that is disjoint from the existing proof, the file is split into three parts to keep the dependencies lighter.

See [zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/basic.20log.20bounds.20on.20harmonic.20sums)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Harmonic/Bounds.lean [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  calc _ = ∫ x in (1)..↑(n+1), x⁻¹ := ?_
+       _ = ∫ x in (1:ℕ)..↑(n+1), x⁻¹ := ?_
+       _ ≤ ∑ d in Finset.Icc 1 n, (d:ℝ)⁻¹ := ?_
+       _ = harmonic n := ?_
+  · rw[integral_inv (by simp[(show ¬ (1:ℝ) ≤ 0 by norm_num)] )]; congr; ring
+  · congr; norm_num
+  · apply AntitoneOn.integral_le_sum_Ico (by norm_num)
+    apply inv_antitoneOn_Icc_right
+    norm_num
+  · simp [harmonic_eq_sum_Icc]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Bounds.lean:30]:
```suggestion
  calc _ = ∫ x in (1:ℕ)..↑(n+1), x⁻¹ := ?_
       _ ≤ ∑ d in Finset.Icc 1 n, (d:ℝ)⁻¹ := ?_
       _ = harmonic n := ?_
  · rw [Nat.cast_one, integral_inv (by simp [(show ¬ (1 : ℝ) ≤ 0 by norm_num)]), div_one]
  · exact (inv_antitoneOn_Icc_right <| by norm_num).integral_le_sum_Ico (Nat.le_add_left 1 n)
  · simp only [harmonic_eq_sum_Icc, Rat.cast_sum, Rat.cast_inv, Rat.cast_coe_nat]
```

File: Mathlib/NumberTheory/Harmonic/Bounds.lean [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  · congr; norm_num;
+  · apply @AntitoneOn.sum_le_integral_Ico 2 (n + 1) fun x : ℝ => (x - 1)⁻¹
+    · linarith [hn]
+    apply sub_inv_antitoneOn_Icc_right; norm_num
+  · convert intervalIntegral.integral_comp_sub_right _ 1
+    · norm_num
+    · simp only [Nat.cast_add, Nat.cast_one, add_sub_cancel]
+  convert integral_inv _
+  · simp
+  norm_num; simp[hn, show (0:ℝ) < 1 by norm_num]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Bounds.lean:56]:
```suggestion
  rw [← Finset.sum_erase_add (Finset.Icc 1 n) _ (Finset.left_mem_Icc.mpr hn), add_comm,
    Nat.cast_one, inv_one]
  refine add_le_add_left ?_ 1
  simp only [Nat.lt_one_iff, Finset.mem_Icc, Finset.Icc_erase_left]
  calc ∑ d : ℕ in .Ico 2 (n + 1), (d : ℝ)⁻¹
    _ = ∑ d in .Ico 2 (n + 1), (↑(d + 1) - 1)⁻¹ := ?_
    _ ≤ ∫ x in (2).. ↑(n + 1), (x - 1)⁻¹  := ?_
    _ = ∫ x in (1)..n, x⁻¹ := ?_
    _ = Real.log ↑n := ?_
  · simp_rw [Nat.cast_add, Nat.cast_one, add_sub_cancel]
  · exact @AntitoneOn.sum_le_integral_Ico 2 (n + 1) (fun x : ℝ ↦ (x - 1)⁻¹) (by linarith [hn]) <|
      sub_inv_antitoneOn_Icc_right (by norm_num)
  · convert intervalIntegral.integral_comp_sub_right _ 1
    · norm_num
    · simp only [Nat.cast_add, Nat.cast_one, add_sub_cancel]
  · convert integral_inv _
    · rw [div_one]
    · simp only [Nat.one_le_cast, hn, Set.uIcc_of_le, Set.mem_Icc, Nat.cast_nonneg,
        and_true, not_le, zero_lt_one]
```

File: Mathlib/NumberTheory/Harmonic/Bounds.lean [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    Real.log y ≤ harmonic ⌊y⌋₊ := by
+  by_cases h0 : y = 0
+  · simp [h0]
+  · calc _ ≤ Real.log ↑(Nat.floor y + 1) := ?_
+        _ ≤ _ := ?_
+    · gcongr
+      apply (Nat.le_ceil y).trans
+      norm_cast
+      exact Nat.ceil_le_floor_add_one y
+    · apply log_add_one_le_harmonic
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Bounds.lean:68]:
```suggestion
  · calc
      _ ≤ Real.log ↑(Nat.floor y + 1) := ?_
      _ ≤ _ := log_add_one_le_harmonic _
    gcongr
    apply (Nat.le_ceil y).trans
    norm_cast
    exact Nat.ceil_le_floor_add_one y
```

File: Mathlib/NumberTheory/Harmonic/Bounds.lean [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+theorem harmonic_floor_le_one_add_log (y : ℝ) (hy : 1 ≤ y) :
+    harmonic ⌊y⌋₊ ≤ 1 + Real.log y := by
+  trans (1 + Real.log (⌊y⌋₊))
+  · apply harmonic_le_one_add_log (⌊y⌋₊)
+  gcongr
+  · norm_cast; apply Nat.lt_of_succ_le; apply Nat.le_floor; norm_cast
+  · apply Nat.floor_le; linarith
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Bounds.lean:76]:
```suggestion
  refine (harmonic_le_one_add_log _).trans ?_
  gcongr
  · exact_mod_cast Nat.floor_pos.mpr hy
  · exact Nat.floor_le <| zero_le_one.trans hy
```

File: Mathlib/NumberTheory/Harmonic/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/-- The nth-harmonic number defined as a finset sum of consecutive reciprocals. -/
+def harmonic : ℕ → ℚ := fun n => ∑ i in Finset.range n, (↑(i + 1))⁻¹
+
+@[simp]
+lemma harmonic_zero : harmonic 0 = 0 :=
+  rfl
+
+@[simp]
+lemma harmonic_succ (n : ℕ) : harmonic (n + 1) = harmonic n + (↑(n + 1))⁻¹ := by
+  apply Finset.sum_range_succ
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Defs.lean:32]:
```suggestion
lemma harmonic_succ (n : ℕ) : harmonic (n + 1) = harmonic n + (↑(n + 1))⁻¹ :=
  Finset.sum_range_succ ..
```

File: Mathlib/NumberTheory/Harmonic/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+lemma harmonic_pos {n : ℕ} (Hn : n ≠ 0) : 0 < harmonic n :=
+  Finset.sum_pos (fun _ _ => inv_pos.mpr (by norm_cast; linarith))
+  (by rwa [Finset.nonempty_range_iff])
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Defs.lean:36]:
```suggestion
  Finset.sum_pos (fun _ _ => inv_pos.mpr (by norm_cast; linarith)) <|
    Finset.nonempty_range_iff.mpr Hn
```

File: Mathlib/NumberTheory/Harmonic/Defs.lean [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+lemma harmonic_eq_sum_Icc {n : ℕ} :  harmonic n = ∑ i in Finset.Icc 1 n, (↑i)⁻¹ := by
+  unfold harmonic
+  rw[Finset.range_eq_Ico, Finset.sum_Ico_add' (fun (i:ℕ) => (i:ℚ)⁻¹) 0 n (c:=1),
+    Nat.Ico_succ_right]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/Harmonic/Defs.lean:41]:
```suggestion
  rw [harmonic, Finset.range_eq_Ico, Finset.sum_Ico_add' (fun (i : ℕ) ↦ (i : ℚ)⁻¹) 0 n (c := 1),
    Nat.Ico_succ_right]
```

**Review by MichaelStollBayreuth — APPROVED** [review]
Modulo some stylistic matters, this looks good to me.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**